src/resolve/resolved-bus.c

```diff
@@ -262,11 +312,15 @@ static void bus_method_resolve_hostname_complete(DnsQuery *query) {
                 goto finish;
 
         r = sd_bus_send(q->manager->bus, reply, NULL);
+        if (r < 0)
+                goto finish;
+
+        q->bus_request = sd_bus_message_unref(q->bus_request);
```

Comment:
Hmm... On transaction failure or any other error paths, only `reply_query_state()` or `reply_method_errno()` are called, and there is no fallback on failure in sending it. I think that should be.

But, when a transaction succeeds, then error on sending message triggers fallback. I forgot the detailed implementation of  sd_bus_send(), but if sd_bus_send() failed after sending message, then two messages are still sent.

I prefer to change the order of `unref` and `send` here, for consistency with other error paths.
```
q->bus_request = sd_bus_message_unref(q->bus_request);
r = sd_bus_send(q->manager->bus, reply, NULL);
```

Reply:
Hmm, I don't understand this comment.

sd_bus_send prepares the message and either writes it to the socket or puts it on the queue. If any of the fails, we should invoke the fallback path in `finish`, including sending of the error message. So I don't think we should unset `q->bus_request` until after sd_bus_send has returned successfully.

Reply:
I mean, for a failed transaction, sd_bus_send() is called only once. But for a successful transaction, sd_bus_send() is called at most twice. Of course, I hope, that happens only when the first send() failed and only one message is sent, but I am not sure.